overlord/snapstate/cookies.go

```diff
@@ -0,0 +1,169 @@
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/strutil"
+)
+
+func cookies(st *state.State) (map[string]string, error) {
+	var snapCookies map[string]string
+	if err := st.Get("snap-cookies", &snapCookies); err != nil {
+		if err != state.ErrNoState {
+			return nil, fmt.Errorf("cannot get snap cookies: %v", err)
+		}
+		snapCookies = make(map[string]string)
+	}
+	return snapCookies, nil
+}
+
+// SyncCookies creates snap cookies for snaps that are missing them (may be the case for snaps installed
+// before the feature of running snapctl outside of hooks was introduced, leading to a warning
+// from snap-confine).
+// It is the caller's responsibility to lock state before calling this function.
+func (m *SnapManager) SyncCookies(st *state.State) error {
+	var snapNames map[string]*json.RawMessage
+	if err := st.Get("snaps", &snapNames); err != nil && err != state.ErrNoState {
+		return err
+	}
+
+	snapCookies, err := cookies(st)
+	if err != nil {
+		return err
+	}
+
+	snapsWithCookies := make(map[string]bool)
+	for _, snap := range snapCookies {
+		if _, ok := snapNames[snap]; !ok || snapsWithCookies[snap] {
+			// there is no point in checking all cookies if we found a bad one - recreate them all
+			snapCookies = make(map[string]string)
+			snapsWithCookies = make(map[string]bool)
+			break
+		}
+		snapsWithCookies[snap] = true
+	}
+
+	var changed bool
+
+	// make sure every snap has a cookie, generate one if necessary
+	for snap := range snapNames {
+		if _, ok := snapsWithCookies[snap]; !ok {
+			cookie := makeCookie()
+			snapCookies[cookie] = snap
+			changed = true
+		}
+	}
+
+	content := make(map[string]*osutil.FileState)
+	for cookie, snap := range snapCookies {
+		content[fmt.Sprintf("snap.%s", snap)] = &osutil.FileState{
+			Content: []byte(cookie),
+			Mode:    0600,
+		}
+	}
+	if _, _, err := osutil.EnsureDirState(dirs.SnapCookieDir, "snap.*", content); err != nil {
+		return fmt.Errorf("Failed to synchronize snap cookies: %s", err)
+	}
+
+	if changed {
+		st.Set("snap-cookies", &snapCookies)
+	}
+	return nil
+}
+
+func (m *SnapManager) createSnapCookie(st *state.State, snapName string) error {
+	snapCookies, err := cookies(st)
+	if err != nil {
+		return err
+	}
+
+	// make sure we don't create cookie if it already exists
+	for _, snap := range snapCookies {
+		if snapName == snap {
+			return nil
+		}
+	}
+
+	cookieID, err := createCookieFile(snapName)
+	if err != nil {
+		return err
+	}
+
+	snapCookies[cookieID] = snapName
+	st.Set("snap-cookies", &snapCookies)
+	return nil
+}
+
+func (m *SnapManager) removeSnapCookie(st *state.State, snapName string) error {
+	var snapCookies map[string]string
+	err := st.Get("snap-cookies", &snapCookies)
+	if err != nil {
+		if err != state.ErrNoState {
+			return fmt.Errorf("cannot get snap cookies: %v", err)
+		}
+		// no cookies in the state
+		if err := removeCookieFile(snapName); err != nil {
```

Comment:
If we can do this if there are no cookies in the state, we could just unconditionally do it at the top of this file, no? I mean, instead of having the same removeCookieFile()  inside two if() just doing it always? Maybe I'm missing something but it slightly bugs me that we cannot have just a single removeCookiesFile() call here in this function.

Reply:
You're right, thanks. Done.